### PR TITLE
[WIP] Add Verse actions hover card

### PR DESCRIPTION
**Pull request by osamasayed**
### Summary
This PR adds ability to show the verse actions menu when hovering over the verse's number

### CODE REVIEW

File: src/utils/verse.ts [hunk 1]
```diff
@@ -70,3 +83,45 @@ export const getWordDataFromLocation = (
...
+/**
+ * Get the verse words. We will put the BE's response for each word
+ * except when the word is the end of the verse then we append some
+ * data that belongs to the verse which will be used when showing the
+ * verse's action menu on hovering over the verse's number.
+ *
+ * @param {Verse} verse
+ * @returns {Word[]}
+ */
+export const getVerseWords = (verse: Verse): Word[] => {
```
Comment (tifa2UP) [inline src/utils/verse.ts:114]: Let's call this `getWordsWithHelperData` or something similar?

File: src/components/Search/SearchResults/SearchResultItem.tsx [hunk 2]
```diff
@@ -17,10 +17,10 @@ const SearchResultItem: React.FC<Props> = ({ result }) => (
         <div className={styles.quranTextContainer}>
           <p className={styles.verseKey}>{result.verseKey}</p>
           <div className={styles.quranTextResult}>
-            {result.words.map((word) => (
+            {result.words.map((word, index) => (
               <QuranWord
                 highlight={!!word.highlight}
-                key={word.id}
+                key={`${result.verseKey}:${index + 1}`}
```
Comment (tifa2UP) [inline src/components/Search/SearchResults/SearchResultItem.tsx:23]: does word.id return a backend related identifier?

File: src/components/Verse/VerseActionsMenu.tsx [hunk 3]
```diff
@@ -20,6 +19,13 @@ import styles from './VerseActionsMenu.module.scss';
 import Modal from '../dls/ModalNew/Modal';
 import VerseAdvancedCopy from './AdvancedCopy/VerseAdvancedCopy';
 
+interface Verse {
+  textUthmani: string;
+  verseKey: string;
+  chapterId: number | string;
+  verseNumber: number;
+}
```
Comment (tifa2UP) [inline src/components/Verse/VerseActionsMenu.tsx:27]: idt this should live in verse Actions, until you name VerseActions :D

File: src/components/Verse/VerseText.tsx [hunk 4]
```diff
@@ -25,14 +25,14 @@ const VerseText = ({ words, isReadingMode = false }: VerseTextProps) => {
     () => isCenterAlignedPage(pageNumber, lineNumber),
     [pageNumber, lineNumber],
   );
-  const firstWordData = getWordDataFromLocation(location);
+  const firstWordData = isFirstWordOfSurah(location);
```
Comment (tifa2UP) [inline src/components/Verse/VerseText.tsx:28]: getFirstWordOfSurah?

File: src/components/dls/QuranWord/QuranWord.tsx [hunk 5]
```diff
@@ -35,31 +39,56 @@ const QuranWord = ({ word, font, highlight, allowWordByWord = true }: QuranWordP
     wordText = <TextWord font={font} text={word.text} charType={word.charTypeName} />;
   }
 
+  const isWordEndOfVerse = isEndOfVerseWord(word);
```
Comment (tifa2UP) [inline src/components/dls/QuranWord/QuranWord.tsx:42]: We can also call this isAyahMarker

File: src/components/dls/QuranWord/QuranWord.tsx [hunk 6]
```diff
@@ -35,31 +39,56 @@ const QuranWord = ({ word, font, highlight, allowWordByWord = true }: QuranWordP
...
-      {showTooltip ? (
-        <Tooltip
-          text={getTooltipText(showTooltipFor, word)}
+      {showVerseActionsMenu ? (
+        <HoverCard
+          openDelay={0}
+          closeDelay={0}
+          body={
+            <VerseActionsMenu
+              verse={{
```
Comment (tifa2UP) [inline src/components/dls/QuranWord/QuranWord.tsx:64]: re: my note above, I wouldn't call this prop verse.

File: src/utils/verse.ts [hunk 7]
```diff
@@ -44,6 +46,17 @@ export const getVerseAndChapterNumbersFromKey = (verseKey: string): [string, str
   return [splits[0], splits[1]];
 };
 
+/**
+ * Split the word's location and get the surahNumber, verseNumber and wordNumber.
+ *
+ * @param {String} wordLocation the word location {surahNumber}:{verseNumber}:{wordNumber}
+ * @returns {[String, String, String]}
+ */
+export const getWordDataByLocation = (wordLocation: string): [string, string, string] => {
```
Comment (tifa2UP) [inline src/utils/verse.ts:55]: getWordLocationByString?

File: src/utils/verse.ts [hunk 8]
```diff
@@ -70,3 +83,45 @@ export const getWordDataFromLocation = (
  * @returns {Verse} verse
  */
 export const getSampleVerse = () => sampleVerse;
+
+/**
+ * Check whether the current word is end of verse the verse based on the char type.
+ * @param {Word} word
+ * @returns {boolean}
+ */
+export const isEndOfVerseWord = (word: Word): boolean => word.charTypeName === CharType.End;
```
Comment (tifa2UP) [inline src/utils/verse.ts:92]: ditto, I prefer ayah marker over end of verse

File: src/components/Search/SearchResults/SearchResultItem.tsx (hunk 2 above)
Comment (osamasayed) [reply src/components/Search/SearchResults/SearchResultItem.tsx:23]: BE just returns: `char_type_name` and `text` for each word without any unique identifiers so have to manually construct the words `location` which is unique.

File: src/components/Verse/VerseActionsMenu.tsx (hunk 3 above)
Comment (osamasayed) [reply src/components/Verse/VerseActionsMenu.tsx]: `Verse` here is a sub-set of `Verse` that only contains the data required for the `VerseActionsMenu` to work.

Comment (osamasayed) [reply src/components/Verse/VerseActionsMenu.tsx]: `Verse` here is a sub-set of `Verse` that only contains the data required for the `VerseActionsMenu` to work. Rename to `ActionsVerse`.

Comment (osamasayed) [reply src/components/Verse/VerseActionsMenu.tsx]: `Verse` here is a sub-set of `Verse` that only contains the data required for the `VerseActionsMenu` to work. Renamed to `ActionsVerse`.

File: src/utils/verse.ts (hunk 7 above)
Comment (naveed-ahmad) [reply src/utils/verse.ts:55]: Or we can get word location from BE as well.

**Pull request closed without merging**